[InstCombine] Prepare foldLogOpOfMaskedICmps to handle trunc to i1. (NFC)

andjo403 opened the pull request:
Making the move of foldLogOpOfMaskedICmps before adding the handling of `trunc to i1` in getMaskedTypeForICmpPair as there was some diffs in llvm-opt-benchmark due to the move. most of the diffs that I looked at was to the better but there was some regressions.

andjo403 commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -199,113 +199,132 @@ static bool decomposeBitTestICmp(Value *LHS, Value *RHS, CmpInst::Predicate &Pre
 
-  // Bail if LHS was a icmp that can't be decomposed into an equality.
-  if (!ICmpInst::isEquality(PredL))
+  } else {
```
will add a "else if" to handled  trunc here in a follow up commit

andjo403 replied:
or maybe better to rearrange the code here to have a decomposeBitTest first that handle the trunc case and the legacy decomposeBitTestIcmp call.
something like `std::optional<DecomposedBitTest> llvm::decomposeBitTest(Value *Cond, bool LookThruTrunc, bool AllowNonZeroC)`

dtcxzyw commented:
>  most of the diffs that I looked at was to the better but there was some regressions.

Can you add some tests for these improvements and regressions?

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -3510,6 +3524,12 @@ Value *InstCombinerImpl::foldBooleanAndOr(Value *LHS, Value *RHS,
+  /// into a single (icmp(A & X) ==/!= Y).
+  if (Value *V = foldLogOpOfMaskedICmps(LHS, RHS, IsAnd, IsLogical, Builder,
+                                        SQ.getWithInstruction(&I)))
+    return V;
```
Maybe keep the original place of the call in this commit so this change can be NFC. Would simplify review.

andjo403 commented:
Changed to only do the preparation to support trunc but do not move the call to a place where any of the values can be trunc so this is NFC now.
Also added decomposeBitTest that only takes a value where the trunc or other bit test can be matched in.

dtcxzyw reviewed (approved):
LGTM. Thank you!

nikic commented on llvm/lib/Analysis/CmpInstAnalysis.cpp:
```diff
@@ -165,3 +165,18 @@ llvm::decomposeBitTestICmp(Value *LHS, Value *RHS, CmpInst::Predicate Pred,
+
+std::optional<DecomposedBitTest>
+llvm::decomposeBitTest(Value *Cond, bool LookThruTrunc, bool AllowNonZeroC) {
+  using namespace PatternMatch;
```
Unused?

andjo403 replied:
yes the using is unused forgot to remove it when I have removed the trunc handling that I hade here. and will add back soon but have removed it now

andjo403 commented:
@nikic do you have any more comments or is this good to go?

andjo403 committed:
[InstCombine] Prepare foldLogOpOfMaskedICmps to handle trunc to i1. (NFC)
